### PR TITLE
Implement for-loop in 0.4

### PR DESCRIPTION
**Pull request by harshdoesdev**
Implement for-loop in 0.4 as mentioned in https://github.com/orgs/fastn-stack/discussions/1079.

**Syntax:**
```
for: $p in $person
```

### CODE REVIEW

File: ftd/src/ast/component.rs [hunk 1]
```diff
@@ -441,51 +447,102 @@ impl Loop {
 
         let loop_statement = loop_header.value.string(doc_id)?;
 
-        let (on, alias) = ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::AS);
+        if loop_header.key.eq("for") {
```
Comment (Arpita-Jaiswal) [inline ftd/src/ast/component.rs:450]:
```suggestion
        if loop_header.key.eq(ftd::ast::utils::FOR) {
```

File: ftd/src/ast/component.rs [hunk 2]
```diff
@@ -441,51 +447,102 @@ impl Loop {
...
+                && !on.starts_with(ftd::ast::utils::CLONE)
+            {
+                return ftd::ast::parse_error(
+                    format!(
+                        "Loop should be on some reference, found: `{}`. Help: use `${}` instead",
+                        on, on
+                    ),
+                    doc_id,
+                    loop_header.line_number,
+                );
```
Comment (Arpita-Jaiswal) [inline ftd/src/ast/component.rs:474]: Copying same logic twice just to get `alias` and `on` differently is not a good idea. Merge the logic and use condition to get `alias` and `on`.

File: ftd/src/ast/component.rs [hunk 3]
```diff
@@ -441,51 +447,102 @@ impl Loop {
...
+        }
     }
 
     fn from_headers(headers: &ftd::p1::Headers, doc_id: &str) -> ftd::ast::Result<Option<Loop>> {
-        let loop_header = headers
-            .0
-            .iter()
-            .find(|v| v.get_key().eq(ftd::ast::utils::LOOP));
+        let loop_header = headers.0.iter().find(|v| {
+            v.get_key().eq(ftd::ast::utils::LOOP) || v.get_key().eq(ftd::ast::utils::FOR)
```
Comment (Arpita-Jaiswal) [inline ftd/src/ast/component.rs:544]:
```suggestion
            [ftd::ast::utils::LOOP, ftd::ast::utils::FOR].contains(v.get_key())
```

File: ftd/src/ast/component.rs [hunk 4]
```diff
@@ -500,44 +557,91 @@ impl Loop {
                 line_number: loop_header.get_line_number(),
             })?;
 
-        let (on, alias) = ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::AS);
+        if loop_header.get_key().eq("for") {
```
Comment (Arpita-Jaiswal) [inline ftd/src/ast/component.rs:560]:
```suggestion
        if loop_header.get_key().eq(ftd::ast::utils::FOR) {
```

**Review by Arpita-Jaiswal — CHANGES_REQUESTED** [review]

File: ftd/src/ast/component.rs [hunk 5]
```diff
@@ -500,44 +557,91 @@ impl Loop {
...
+                return ftd::ast::parse_error(
+                    format!(
+                        "Loop should be on some reference, found: `{}`. Help: use `${}` instead",
+                        on, on
+                    ),
+                    doc_id,
+                    loop_header.get_line_number(),
+                );
+            }
+
```
Comment (Arpita-Jaiswal) [inline ftd/src/ast/component.rs:576]: Again don't copy the logic twice.

**Review by Arpita-Jaiswal — CHANGES_REQUESTED** [review]

**Comment by amitu** [comment]
WE have to ensure both old and new syntax work. Old syntax should issue a warning.

File: ftd/src/ast/component.rs [hunk 6]
```diff
@@ -441,7 +449,55 @@ impl Loop {
...
         let loop_statement = loop_header.value.string(doc_id)?;
 
-        let (on, alias) = ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::AS);
+        let is_for_loop = loop_header.key.eq(ftd::ast::utils::FOR);
+
+        let (alias, on, key_name) = if is_for_loop {
+            let (pair, on) =
+                ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::IN);
+
+            let on = if let Some(on) = on {
```
Comment (Heulitig) [inline ftd/src/ast/component.rs:458]: Change the below unwrap to this 
```suggestion
            let on = on.ok_or(ftd::ast::Error::Parse {
                message: "Statement \"for\" needs a list to operate on".to_string(),
                doc_id: doc_id.to_string(),
                line_number: loop_header.line_number,
            })?;
```

File: ftd/src/ast/component.rs [hunk 7]
```diff
@@ -500,7 +540,55 @@ impl Loop {
...
             })?;
 
-        let (on, alias) = ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::AS);
+        let is_for_loop = loop_header.get_key().eq(ftd::ast::utils::FOR);
+
+        let (alias, on, key_name) = if is_for_loop {
+            let (pair, on) =
+                ftd::ast::utils::split_at(loop_statement.as_str(), ftd::ast::utils::IN);
+
+            let on = if let Some(on) = on {
```
Comment (Heulitig) [inline ftd/src/ast/component.rs:549]: Similarly update this as well

File: ftd/src/ast/component.rs [hunk 8]
```diff
@@ -513,29 +601,14 @@ impl Loop {
...
-            }
-        };
+        let alias = alias
+            .trim_start_matches(ftd::ast::utils::REFERENCE)
+            .to_string();
 
         Ok(Some(Loop::new(
             on.as_str(),
             alias.as_str(),
+            key_name,
```
Comment (Heulitig) [inline ftd/src/ast/component.rs:611]: Rename key_name to loop_counter_alias
```suggestion
            loop_counter_alias,
```

File: ftd/src/interpreter/things/component.rs [hunk 9]
```diff
@@ -1107,15 +1108,22 @@ fn search_things_for_module(
...
+    pub key_name: Option<String>,
     pub line_number: usize,
 }
 
 impl Loop {
-    fn new(on: ftd::interpreter::PropertyValue, alias: &str, line_number: usize) -> Loop {
+    fn new(
+        on: ftd::interpreter::PropertyValue,
+        alias: &str,
+        key_name: Option<String>,
```
Comment (Heulitig) [inline ftd/src/interpreter/things/component.rs:1119]: Similarly refactor this variable name as well
```suggestion
        loop_counter_alias: Option<String>,
```

File: ftd/src/interpreter/utils.rs [hunk 10]
```diff
@@ -281,7 +281,7 @@ pub fn get_argument_for_reference_and_remaining(
             };
         }
     }
-    if let Some((loop_name, loop_argument)) = loop_object_name_and_kind {
+    if let Some((loop_name, loop_argument, loop_key_name)) = loop_object_name_and_kind {
```
Comment (Heulitig) [inline ftd/src/interpreter/utils.rs:284]: Same here
```suggestion
    if let Some((loop_name, loop_argument, loop_counter_alias)) = loop_object_name_and_kind {
```

File: ftd/src/ast/component.rs [hunk 11]
```diff
@@ -441,7 +449,55 @@ impl Loop {
...
+
+            let (alias, key_name) = ftd::ast::utils::split_at(pair.as_str(), ", ");
+
+            (alias, on, key_name)
+        } else {
+            use colored::Colorize;
+
+            println!(
+                "{}",
+                "Warning: \"$loop$\" is deprecated, use \"for:\" instead".bright_yellow()
```
Comment (Heulitig) [inline ftd/src/ast/component.rs:476]: Also fix this warning message typo 
```suggestion
                "Warning: \"$loop$\" is deprecated, use \"for\" instead".bright_yellow()
```

**Review by Heulitig — CHANGES_REQUESTED** [review]

**Review by Heulitig — APPROVED** [review]

**Review by Arpita-Jaiswal — APPROVED** [review]

**Pull request merged**